### ISSUE TITLE
Automatic font expansion

### ISSUE THREAD

**Issue by Enivex** [typst/typst#4693]
### Description

The [microtype](https://ctan.org/pkg/microtype?lang=en) package in LaTeX provides several microtypographical features. The two main ones, enabled by default in pdfTeX and luaTeX, are protrusion and expansion.

Typst already supports a (very) limited form of protrusion, and there's an existing issue concerning improving this https://github.com/typst/typst/issues/261

However, expansion is currently completely absent from Typst, as far as I'm aware.

Automatic font expansion can subtly (default max in microtype is 2%) change the width of a font (through stretching) in order to achieve better typesetting results.

### Use Case

Font expansion can improve improve text regularity and greatly reduce the need for hyphenation.

**Comment by eclecticfluff** [comment]
Note that microtype does font expansion by directly modifying glyphs. While this does work and provides a reasonable fallback, using the width axis in variable fonts that have it would be ideal.

**Comment by Enivex** [comment]
> Note that microtype does font expansion by directly modifying glyphs. While this does work and provides a reasonable fallback, using the width axis in variable fonts that have it would be ideal.

Yes, that would be a better option for variable fonts. However, I do not think all of them have a width axis.

(Also typst does not yet support variable fonts.)

**Comment by corneliusroemer** [comment]
The [microtype user manual](https://mirror.foobar.to/CTAN/macros/latex/contrib/microtype/microtype.pdf) is a good resource for inspiration.

**Comment by Krasjet** [comment]
I'll add some more resources in case someone wants to take on the glyph expansion part:
1. The extension to the Knuth-Plass algorithm used by pdftex is explained in this [thesis](https://www.pragma-ade.nl/pdftex/thesis.pdf). (the implementation in pdftex is a [literal program](https://texdoc.org/serve/pdftex/1) as well)
2. There was a demo made by @ammonj for justification using variable font glyph expansion: https://ammonj.github.io/better-justification-demo/. It's not using Knuth-Plass, but the idea is there.

**Comment by omentic** [comment]
Support for variable fonts has just been merged (https://github.com/typst/typst/pull/8425) 🎉
